integration_tests/run_pyspark_from_build.sh

```diff
@@ -70,6 +70,30 @@ else
         # to still launch a spark application.
         TEST_PARALLEL=`nvidia-smi --query-gpu=memory.free --format=csv,noheader | awk '{if (MAX < $1){ MAX = $1}} END {print int(MAX / (2.3 * 1024)) - 1}'`
         echo "AUTO DETECTED PARALLELISM OF $TEST_PARALLEL"
+
+        # adjust TEST_PARALLEL according to cpu cores and free memory;
+        # first read from ENV; Note it's hardly to get CPU cores and real free memory in docker
+        if [[ "${PYSP_TEST_CPU_CORES}" != "" ]]; then
+          cpu_cores=${PYSP_TEST_CPU_CORES}
+        else
+          cpu_cores=$(nproc --all)
+        fi
+        if [[ "${PYSP_TEST_FREE_MEMORY}" != "" ]]; then
+          free_mem_mib=${PYSP_TEST_FREE_MEMORY}
+        else
+          free_mem_mib=$(awk '/MemFree/ { printf "%d\n", $2/1024 }' /proc/meminfo)
+        fi
+        max_parallel_for_cpu_cores=$((${cpu_cores} - 1))
```

Comment:
Adding a comment for why the `- 1` is here would help future maintainers.  Also do we want to subtract 1 even if an environment variable explicitly told us how many to use?

Reply:
Added comment.  IIUC environment variable tells us the number of all available cores, but we do not want to use all of them.